### PR TITLE
Join Implementation

### PR DESCRIPTION
**Pull request by omermadmon**
A sort merge join implementation utilizing Datastore API. Will be used for retrieving all statistics entities corresponding to all simulations matching a certain condition (filter).

### CODE REVIEW

File: src/main/java/com/google/research/bleth/utils/Queries.java [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+
+        // If one of these queries is empty, return an empty list.
+        if (!(primaryKeyIterator.hasNext() && secondaryEntityIterator.hasNext())) {
+            return result;
+        }
+
+        // Iterate over two queries results and add secondary entities with a matching foreign key.
+        String primaryKey = primaryKeyIterator.next();
+        Entity secondaryEntity = secondaryEntityIterator.next();
+        String secondaryKey = String.valueOf(secondaryEntity.getProperty(foreignKey));
```
Comment (nolaviz) [inline src/main/java/com/google/research/bleth/utils/Queries.java:63]: Consider adding a helper method that does both of these steps together, so you don't need to duplicate them multiple times.

(Even better, introduce a second helper - with a similar name - that just wraps the `primaryKeyIterator.next()` call: then the code _here_ to advance both iterators would look the same.)

File: src/main/java/com/google/research/bleth/utils/Queries.java [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+        while (primaryKeyIterator.hasNext() && secondaryEntityIterator.hasNext()) {
+            if (primaryKey.compareTo(secondaryKey) > 0) {
+                // primaryKey > secondaryKey
```
Comment (nolaviz) [inline src/main/java/com/google/research/bleth/utils/Queries.java:66]: This comment isn't really needed: any sufficiently-experienced Java programmer knows the code in line 65 is the right way to write the condition in line 66.

**Review by nolaviz — APPROVED** [review]
Nice!

File: src/main/java/com/google/research/bleth/utils/Queries.java (hunk 1 above)
Comment (omermadmon) [reply src/main/java/com/google/research/bleth/utils/Queries.java:63]: I will leave it as is because of lines 74 and 83, as discussed.

**Pull request merged**